velox/type/Filter.h

```diff
@@ -2162,6 +2163,55 @@ static inline bool applyFilter(TFilter& filter, StringView value) {
   return filter.testBytes(value.data(), value.size());
 }
 
+namespace detail {
+
+template <typename TInput, typename TOutput>
+TOutput rescale(
+    TInput value,
+    uint8_t fromPrecision,
+    uint8_t fromScale,
+    uint8_t toPrecision,
+    uint8_t toScale) {
+  TOutput rescaled;
+  const auto status = DecimalUtil::rescaleWithRoundUp<TInput, TOutput>(
+      value, fromPrecision, fromScale, toPrecision, toScale, rescaled);
+  VELOX_CHECK(status.ok(), status.message());
+  return rescaled;
+}
+
+} // namespace detail
+
+/// Helper for applying filters with type evolution. For decimal value, if the
+/// filter and value types are different, the value is rescaled to the filter
+/// type.
+template <typename TFilter, typename T>
+static inline bool applyTypeEvolutionFilter(
+    TFilter& filter,
+    T value,
+    const TypePtr& filterType,
+    const TypePtr& valueType) {
+  VELOX_CHECK(filterType && valueType);
+  if (filterType->equivalent(*valueType)) {
```

Comment:
We cannot afford to do this check for each row for all types.  Can we only do this when we know the column is decimal type?  Other types do not need rescale.